### PR TITLE
add where_are_my_files.txt

### PR DESCRIPTION
**Pull request by phue**
Adresses #42

This implements the same feature present in nf-core/rnaseq , informing users about missing intermediate files. 

I also found a bug when running the pipeline with the `--unmapped` flag:
```
ERROR ~ No such variable: bismark_unmapped
```

This could be fixed by exchanging the if condition in the output declaration with nextflow's [`optional true`](https://www.nextflow.io/docs/latest/process.html#optional-output) attribute

### CODE REVIEW

**Comment by phue** [comment]
Thanks for your review, I changed it accordingly for all file() statements. While I was at it made use of the new `checkIfExists` attribute.

BTW, it seems you missed one your patch for the rnaseq pipeline:

https://github.com/nf-core/rnaseq/blob/565a8a5e216d528c18c74b59ecee29747e6eaab8/main.nf#L161

File: main.nf [hunk 1]
```diff
@@ -259,11 +266,11 @@ if(!params.bwa_meth_index && params.fasta && params.aligner == 'bwameth'){
  */
 if(!params.fasta_index && params.fasta && params.aligner == 'bwameth'){
     process makeFastaIndex {
-        tag fasta
+        tag "$fasta"
         publishDir path: "${params.outdir}/reference_genome", saveAs: { params.saveReference ? it : null }, mode: 'copy'
 
         input:
-        file fasta
+        file fasta from fasta_2
```
Comment (apeltzer) [inline main.nf]: Can you please use names that are more explicit? 

`ch_from_blabla_to_blabla` is something that helps you (and me, others) to keep things traceable :-)

File: main.nf [hunk 2]
```diff
@@ -237,11 +244,11 @@ if(!params.bismark_index && params.fasta && params.aligner == 'bismark'){
  */
 if(!params.bwa_meth_index && params.fasta && params.aligner == 'bwameth'){
     process makeBwaMemIndex {
-        tag fasta
+        tag "$fasta"
         publishDir path: "${params.outdir}/reference_genome", saveAs: { params.saveReference ? it : null }, mode: 'copy'
 
         input:
-        file fasta from fasta
+        file fasta from fasta_1
```
Comment (apeltzer) [inline main.nf]: Can you please use names that are more explicit?

ch_from_blabla_to_blabla is something that helps you (and me, others) to keep things traceable :-)

File: main.nf [hunk 3]
```diff
@@ -217,7 +224,7 @@ if(!params.bismark_index && params.fasta && params.aligner == 'bismark'){
                    saveAs: { params.saveReference ? it : null }, mode: 'copy'
 
         input:
-        file fasta from fasta
+        file fasta from fasta_1
```
Comment (apeltzer) [inline main.nf]: Can you please use names that are more explicit?

ch_from_blabla_to_blabla is something that helps you (and me, others) to keep things traceable :-)

File: main.nf [hunk 4]
```diff
@@ -32,30 +32,37 @@ if (params.aligner != 'bismark' && params.aligner != 'bwameth'){
...
 else if( !params.fasta ) {
     exit 1, "No reference genome index specified!"
 }
+
 if ( params.fasta ){
-    fasta = file(params.fasta)
-    if( !fasta.exists() ) exit 1, "Fasta file not found: ${params.fasta}"
+    Channel
+        .fromPath(params.fasta, checkIfExists: true)
+        .into { fasta_1; fasta_2; fasta_3 }
```
Comment (apeltzer) [inline main.nf]: Can you please use names that are more explicit?

`ch_fasta_for_XYZ` is something that helps you (and me, others) to keep things traceable :-)

File: main.nf [hunk 5]
```diff
@@ -316,7 +323,7 @@ if(params.notrim){
 
         input:
         set val(name), file(reads) from read_files_trimming
-        file wherearemyfiles
+        file wherearemyfiles from wherearemyfiles_1
```
Comment (apeltzer) [inline main.nf]: Same here, please use explicit names for channels :-) 
It gets super confusing at some point otherwise....

File: main.nf [hunk 6]
```diff
@@ -360,7 +367,7 @@ if(params.aligner == 'bismark'){
         input:
         set val(name), file(reads) from trimmed_reads
         file index from bismark_index.collect()
-        file wherearemyfiles
+        file wherearemyfiles from wherearemyfiles_2
```
Comment (apeltzer) [inline main.nf]: Same here, please use explicit names for channels :-)
It gets super confusing at some point otherwise....

File: main.nf [hunk 7]
```diff
@@ -595,7 +602,7 @@ if(params.aligner == 'bwameth'){
         input:
         set val(name), file(reads) from trimmed_reads
         file bwa_meth_indices from bwa_meth_indices.collect()
-        file wherearemyfiles
+        file wherearemyfiles from wherearemyfiles_2
```
Comment (apeltzer) [inline main.nf]: Same here, please use explicit names for channels :-)
It gets super confusing at some point otherwise....

File: main.nf [hunk 8]
```diff
@@ -795,7 +802,7 @@ process multiqc {
     publishDir "${params.outdir}/MultiQC", mode: 'copy'
 
     input:
-    file multiqc_config
+    file multiqc_config from multiqc_config
```
Comment (apeltzer) [inline main.nf]: This one is fine :-)

**Comment by phue** [comment]
I decided not to follow the ch_* naming scheme because that somehow introduces inconsistencies, as the numbering approach has been used throughout this pipeline. See for example:

https://github.com/nf-core/methylseq/blob/93600734bc7a1c1a12a3848b4429425ee3aee79c/main.nf#L372

https://github.com/nf-core/methylseq/blob/93600734bc7a1c1a12a3848b4429425ee3aee79c/main.nf#L474

While I agree having a 'ch_' prefix would be nice, I think would be out of scope for this PR. It could be done in another PR though, but then for all Channels (for the sake of consistency).

**Comment by phue** [comment]
https://speakerdeck.com/pditommaso/the-state-of-nextflow?slide=15

This will solve the problem anyway :)

**Comment by apeltzer** [comment]
Well, I guess its something we should definitely address this in general ;-)

**Comment by ewels** [comment]
I'm slightly skeptical about not fixing a problem because the same problem exists elsewhere, but I see where you're coming from. I think this is fine to merge now, but let's address the crappy channel naming soon if we can. (Apologies, I'm responsible for the majority of this).

**Comment by ewels** [comment]
I'm slightly dubious that the `checkIfExists` flag works in Nextflow 0.32.0 (the minimum version supported). Whilst looking I realised that the `.travis.yml` config has an error which means it doesn't use the existing references as intended. I'll fix this in another PR though 👍

**Comment by ewels** [comment]
Travis fix addressed in #55

**Comment by ewels** [comment]
Also tracked down when `checkIfExists` was released (see [comment](https://github.com/nextflow-io/nextflow/issues/874#issuecomment-432715504)). So I'm expecting tests to fail with nextflow version under `18.10.1`.

**Comment by ewels** [comment]
I take it back - tests seem happy about the channel change! Perhaps older versions of nextflow simply ignore the unrecognised flag? Anyway, if the tests are happy then I'm happy...

**Comment by phue** [comment]
> Also tracked down when `checkIfExists` was released (see [comment](https://github.com/nextflow-io/nextflow/issues/874#issuecomment-432715504)). So I'm expecting tests to fail with nextflow version under `18.10.1`.

The flag itself was introduced with this commit :  
https://github.com/nextflow-io/nextflow/commit/8bc6c99824c22b6a0e94ecf1f0e61267c3176c25
(released with https://github.com/nextflow-io/nextflow/releases/tag/v0.32.0)

I guess that is why the tests don't fail, but with 0.32.0 there would be an edge case if glob patterns don't match any files. Maybe it would be wise to bump the required version to 18.10.1

> I'm slightly skeptical about not fixing a problem because the same problem exists elsewhere, but I see where you're coming from. 

I fully agree on that, but I think its a design choice to be made. As @apeltzer pointed out, it's much easier to trace the fate of a channel and read a pipeline if the name indicates where it comes from and where it goes, but on the other hand, once the .fork() operator is released it might as well make sense to just indicate what the channel contains, because it can only be consumed as many times as it is forked.
